src/mast/mod.rs

```diff
@@ -0,0 +1,1108 @@
+            self.vars.remove(&d);
+        }
+    }
+
+    /// Returns true if a scope is a prefix of our scope.
+    pub fn is_in_scope(&self, prefix_scope: usize) -> bool {
+        self.current_scope >= prefix_scope
+    }
+
+    /// Stores type information about a local variable.
+    pub fn store_type(&mut self, ident: &str, type_info: &MTypeInfo) -> Result<()> {
+        match self
+            .vars
+            .insert(ident.to_string(), (self.current_scope, type_info.clone()))
+        {
+            Some(_) => Err(Error::new(
+                "type-checker",
+                ErrorKind::DuplicateDefinition(ident.to_string()),
+                type_info.span,
+            )),
+            None => Ok(()),
+        }
+    }
+
+    /// Retrieves type information on a variable, given a name.
+    /// If the variable is not in scope, return None.
+    pub fn get_type_info(&self, ident: &str) -> Option<&MTypeInfo> {
+        if let Some((scope, type_info)) = self.vars.get(ident) {
+            if self.is_in_scope(*scope) {
+                Some(type_info)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// A context to store the last node id for the monomorphized AST.
+#[derive(Debug, Default)]
+pub struct MastCtx {
+    pub last_node_id: usize,
+}
+
+impl MastCtx {
+    pub fn new() -> Self {
+        Self { last_node_id: 0 }
+    }
+
+    pub fn next_node_id(&mut self) -> usize {
+        self.last_node_id += 1;
+        self.last_node_id
+    }
+}
+
+#[derive(Debug)]
+/// Mast relies on the TAST for the information about the "unresolved" types to monomorphize.
+/// Things such as loading the function AST and struct AST from fully qualified names.
+/// After monomorphization process, the following data will be updated:
+/// - Resolved types. This can be used to determine the size of a type.
+/// - Instantiated functions. The circuit writer will load the instantiated function AST by node id.
+/// - Monomorphized AST is generated for the circuit writer to walk through and compute.
+pub struct Mast<B>
+where
+    B: Backend,
+{
+    tast: TypeChecker<B>,
+
+    /// Mapping from node id to resolved type
+    node_types: HashMap<usize, TyKind>,
+
+    /// Mapping from node id to instantiated functions
+    node_functions: HashMap<usize, FnInfo<B>>,
+
+    ctx: MastCtx,
+
+    /// Monomorphized AST of the main function
+    pub main_fn_ast: Option<FunctionDef>,
+}
+
+impl<B: Backend> Mast<B> {
+    pub fn new(tast: TypeChecker<B>) -> Self {
+        Self {
+            tast,
+            node_types: HashMap::new(),
+            node_functions: HashMap::new(),
+            ctx: MastCtx::new(),
+            main_fn_ast: None,
+        }
+    }
+
+    pub fn expr_type(&self, expr: &Expr) -> Option<&TyKind> {
+        self.node_types.get(&expr.node_id)
+    }
+
+    pub fn expr_fn(&self, expr: &Expr) -> Option<&FnInfo<B>> {
+        self.node_functions.get(&expr.node_id)
```

Comment:
actually, the `node_functions` stores all the functions including non-generic functions. I think it would be better to store functions by function signature instead as described above.

Reply:
we will have another PR to refactor to based on function signature.